Rewriting image augmenter

chsasank opened the pull request:
Hi,

As discussed on #2394, I rewrote `ImageDataGenerator` using homographies. There's a significant speedup compared to the original implementation.

Minor downside is that scikit-image is now a dependency.

joelthchao commented:
I think we should keep the old one and create new class `SkimageDataGenerator(ImageDataGenerator)` for better support.

chsasank commented:
Actually, the old one is buggy.
I ran a small [script](https://gist.github.com/chsasank/2f392206b07b70a298b93f518bdb5b29) and saved images generated by the augmenter.

Original image (`random_img.jpg`):
![random_img](https://cloud.githubusercontent.com/assets/9305875/14707573/0e3a493c-07e3-11e6-8fae-dfc87f4dbea4.jpg)


These are generated by keras:

![gen_0_keras](https://cloud.githubusercontent.com/assets/9305875/14707552/01a47850-07e3-11e6-8faf-9e1275a660e6.jpg)
![gen_1_keras](https://cloud.githubusercontent.com/assets/9305875/14707556/01f7f4d0-07e3-11e6-8d4b-20c10ae8be7f.jpg)
![gen_2_keras](https://cloud.githubusercontent.com/assets/9305875/14707555/01f798b4-07e3-11e6-979b-5341a8cbf7bd.jpg)
![gen_3_keras](https://cloud.githubusercontent.com/assets/9305875/14707554/01f69996-07e3-11e6-9e06-b9abc1b330f6.jpg)
![gen_4_keras](https://cloud.githubusercontent.com/assets/9305875/14707557/01f9f8d4-07e3-11e6-8404-90c00d88e408.jpg)
 
And these by the fork:
![gen_0_fork](https://cloud.githubusercontent.com/assets/9305875/14707590/25ca0f74-07e3-11e6-9b3a-d795befe8085.jpg)
![gen_1_fork](https://cloud.githubusercontent.com/assets/9305875/14707592/261a13f2-07e3-11e6-8727-9ba22a447ba6.jpg)
![gen_2_fork](https://cloud.githubusercontent.com/assets/9305875/14707591/2618aa26-07e3-11e6-9115-69a8986506ec.jpg)
![gen_3_fork](https://cloud.githubusercontent.com/assets/9305875/14707594/261c86e6-07e3-11e6-82e6-7222c195daa6.jpg)
![gen_4_fork](https://cloud.githubusercontent.com/assets/9305875/14707593/261bfdca-07e3-11e6-9b24-44361a3ebbb3.jpg)

So, this warrants the removal of old `ImageDataGenerator`

albertomontesg commented:
I think that the transformation methods shouldn't be removed because it would be useful to use them at writing your own DataGenerator with custom data augmentation.

joelthchao commented:
@albertomontesg Thanks, these useful methods should be kept for the future use.

chsasank commented:
> I think that the transformation methods shouldn't be removed because it would be useful to use them at writing your own DataGenerator with custom data augmentation.

But people can do the same with the new one, right?
All that has to be changed is `random_transform`

joelthchao commented:
@chsasank Better way is to make functions which only do a single transformation. By the way, I fix several bugs in `fit` and I am going to temporally keep those codes and wait for next refactor to replace them.

chsasank commented:
@joelthchao  
>Better way is to make functions which only do a single transformation

But this defeats the purpose of this PR right?

What bugs were there in `fit`?

joelthchao commented:
@chsasank `fit` convert numpy array to PIL Image and send to `random_transform`. Your code assume input is in numpy type, which is right during `flow` but fail in `fit` when it calls `x.dtype`.

chsasank commented:
Cool, I get what you mean by keeping transformation methods intact.
I'll rewrite these functions to use skimage instead of ndimage.

fchollet commented:
I don't think we should depend on skimage at all. Any way to get the same speed-up while only relying on Numpy/Scipy?

dansbecker commented:
Wow.  That issue with the colors seems bad.  If you guys want someone to look into that, I'm happy to do so.  If we're pulling all that code out anyway, then I won't spend time on it.

chsasank commented:
Hi @fchollet ,
I think we can use [this](http://docs.scipy.org/doc/scipy-0.16.0/reference/generated/scipy.ndimage.interpolation.affine_transform.html#scipy.ndimage.interpolation.affine_transform) instead of skimage. 
I'll test and make changes.

fchollet commented:
> Wow. That issue with the colors seems bad.

I'm skeptical this is for real, I remember generating test images when this code was initially written and they looked fine. I'm 95% sure that OP is either normalizing the arrays in some way, or messing up the image saving step.

fchollet commented:
> I think we can use this instead of skimage. 

Cool, that sounds good.

chsasank commented:
I have rewritten using scipy's ndimage.
I think the bug in the previous code was here [1](https://github.com/fchollet/keras/blob/master/keras/preprocessing/image.py#L37) and [2](https://github.com/fchollet/keras/blob/master/keras/preprocessing/image.py#L61).

ndimage treats every dimension in an array as spatial dimension. See [this](http://stackoverflow.com/a/9284454/6212301).

So the operation [1] would work only if channel axis = 0 while [2] would work only if channel axis = 2.
I have fixed this.

chsasank commented:
BTW, can somebody explain me what `random_channel_shift` is?
There's a `todo` in the code.

I will try to implement it.

fchollet commented:
> ndimage treats every dimension in an array as spatial dimension.

Right, I get it now.

> BTW, can somebody explain me what random_channel_shift is?

It's a random shift in RGB color channels, effectively changing the color scheme of a picture. e.g. +10 on the R channel, -20 on the G channel, +5 on the B channel...

chsasank commented:
Is this like what alexnet paper does?

From the paper:
>The second form of data augmentation consists of altering the intensities of the RGB channels in training images. Specifically, we perform PCA on the set of RGB pixel values throughout the ImageNet training set. To each training image, we add multiples of the found principal components, with magnitudes proportional to the corresponding eigenvalues times a random variable drawn from
a Gaussian with mean zero and standard deviation 0.1

fchollet commented:
> Is this like what alexnet paper does?

It may be overkill, and it is less practical because it requires fitting on training data, but yes, that would be a good strategy.

chsasank commented:
We can have pertained components trained on imagenet as an option if an user doesn't want to fit on his own data. Is this a good idea?

chsasank commented:
Also, `featurewise_centre` and `featurewise_centre`  do not do simple normalisation of channels but they have mean image and stddev image. Is this intended?

joelthchao commented:
@chsasank 
> We can have pertained components trained on imagenet as an option if an user doesn't want to fit on his own data. Is this a good idea?

I think it would make this component too complicated to use, probably you can make a comprehensive example to illustrate how to do proper data augmentation based on Imagenet

> Also, featurewise_centre and featurewise_centre do not do simple normalisation of channels but they have mean image and stddev image. Is this intended?

It fits to train data, which is a standard normalization method in computer vision.

Also, you can add `random_zoom` to TODO or I can implement it now.

fchollet commented:
> Also, featurewise_centre and featurewise_centre do not do simple normalisation of channels but they have mean image and stddev image. Is this intended?

Channelwise normalization would be a useful addition as well. I think it would be fine to make that mode work on separate channels differently rather than globally on all channels.

Please let me know when this is ready for merging.

joelthchao commented:
Currently, `standardize` do the following things:
1. `samplewise`, normalize over channel for a single image
2. `featurewise`: substract mean image and divide std image of data set, which is channel independent.
@fchollet do you mean creating a `channelwise` mode? 
PS. most of works only do `featurewise` normalization

chsasank commented on keras/preprocessing/image.py:
```diff
@@ -70,9 +70,13 @@ def random_barrel_transform(x, intensity):
-    pass
+def random_channel_shift(x, intensity, channel_index=0):
+    x = np.rollaxis(x, channel_index, 0)
+    channel_images = [np.clip(x_channel + np.random.uniform(-intensity, intensity), 0, 1) 
```
Why clip to [0,1]?
There's no guarantee that X passed to the datagenerator will be within 0 and 1.

chsasank commented on keras/preprocessing/image.py:
```diff
@@ -6,105 +6,137 @@
-                                               cval=cval)
+def random_channel_shift(x, intensity, channel_index=0):
+    x = np.rollaxis(x, channel_index, 0)
+    channel_images = [np.clip(x_channel + np.random.uniform(-intensity, intensity), 0, 1) 
```
Why clip to [0,1]
There's no guarantee that X passed to data generator will be within 0 and 1.
For example uint8 images can be passed to data generator.

joelthchao replied:
Got it! Let me check if other codes make such guarantee.

joelthchao replied:
array_to_img need argument scale to decide if it is needed to scale to uint8. And the previous code assume X is always within 0 and 1. We may need to fix it.

chsasank replied:
>And the previous code assume X is always within 0 and 1. We may need to fix it.

No all the random transformations do not assume anything about the scale. I have checked it passing uint8 images.

Also, `array_to_img` has an argument `scale` to rescale the images to [0,256] for writing purposes.

chsasank replied:
For `random_channel_shift` we can change the offending line to 
```
channel_images = [np.clip(x_channel + np.random.uniform(-intensity, intensity), np.min(x), np.max(x)) 
                  for x_channel in x]
```

joelthchao commented:
@fchollet I think it is ready. @chsasank can you double check the codes?
Barrel/Fisheye effect may need cv2 library support, so I will leave them along.

chsasank commented on docs/templates/preprocessing/image.md:
```diff
@@ -28,6 +28,10 @@ Generate batches of tensor image data with real-time data augmentation. The data
     - __width_shift_range__: Float (fraction of total width). Range for random horizontal shifts.
     - __height_shift_range__: Float (fraction of total height). Range for random vertical shifts.
     - __shear_range__: Float. Shear Intensity (Shear angle in counter-clockwise direction as radians)
+    - __zoom_range__: Float or [lower, upper]. Range for random zoom.
```
Shouldn't we explain if zoom_range is float `[lower, upper]` is `[1-zoom_range, 1+ zoom_range]`, because it is not apparent from the documentation.

chsasank commented:
>Barrel/Fisheye effect may need cv2 library support, so I will leave them alone.

We might be able to implement it with `ndimages`, but it might be slow. See [this](http://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.ndimage.interpolation.map_coordinates.html).
We can leave this for a future PR.

chsasank commented on keras/preprocessing/image.py:
```diff
@@ -6,105 +6,137 @@
+def random_channel_shift(x, intensity, channel_index=0):
+    x = np.rollaxis(x, channel_index, 0)
+    min_x, max_x = np.min(x), np.max(x)
+    channel_images = [np.clip(x_channel + np.random.uniform(-intensity, intensity), min_x, max_x)
```
On a second thought, for better UX shouldn't we ask user to specify channel shift `intensity` in terms of fraction of  dynamic range, instead of absolute values? (like in `height_shift_range `)

joelthchao replied:
If fraction is used for channel shift `intensity` then we might need to know whether user is using `uint8` or `float32`. But can we know that?

chsasank replied:
We can just use `min_x, max_x = np.min(x), np.max(x)` for dynamic range needs right?

joelthchao replied:
That is not 100% guarantee and it will result in different intensity for different image.

chsasank replied:
True, it might be slightly off [0,255] for uint8 image or [0,1] for a float image, but it usually is a very good approximation. 
>it will result in different intensity for different image.

We're adding random intensities anyway, right?

joelthchao replied:
For example, given 20% intensity shift, an image with pixel value [50, 100] will have larger shift than an image with pixel value in [70, 80].

chsasank replied:
Of course. But natural images usually use almost complete dynamic range. Think about it, in a uint8 64x64x3 image, there are about 12k pixels. There's a very high probability min is 0 and max is 255 or 254. 

It's just a suggestion. We can keep current one as it is, if you think that is better.

fchollet commented:
> @fchollet I think it is ready.

Cool, will review later today. Thanks!

fchollet commented on keras/preprocessing/image.py:
```diff
@@ -6,105 +6,137 @@
 
+def random_rotation(x, rg, row_index=1, col_index=2, channel_index=0,
+                    fill_mode='nearest', cval=0.):
+    theta = np.pi/180*np.random.uniform(-rg, rg)
```
Style nitpick, but please insert spaces around operators (`*`, `+`, etc).

fchollet commented on keras/preprocessing/image.py:
```diff
@@ -6,105 +6,137 @@
 
+def random_zoom(x, zoom_range, row_index=1, col_index=2, channel_index=0,
+                fill_mode='nearest', cval=0.):
+    if zoom_range != [1., 1.]:
```
I think we should expect a list, but also allow for tuples. In which case this comparison won't work; better to compare each member: `zoom_range[0] == 1 and zoom_range[1] == 1` .

Also, since we make the assumption that `zoom_range` is going to be a 2-element list/tuple, better check first and raise a helpful error message if it's not the case.

fchollet commented on keras/preprocessing/image.py:
```diff
@@ -6,105 +6,137 @@
     return x
 
+def transform_matrix_offset_center(matrix, x, y):
+    o_x = float(x)/2 + 0.5
```
Spaces around `/`

fchollet commented:
Cool, looks good to me. Just a few minor nitpicks.

fchollet commented:
Thank you for this useful contribution, merging now.
